### PR TITLE
Riot replaces DOMstring append when rendering messages

### PR DESCRIPTION
**Pull request by rjmk**
Hi @dwyl

As discussed in https://github.com/dwyl/hapi-socketio-redis-chat-example/issues/20, this PR adds riot and has a series of commits that form a step-by-step guide.

Also, solves #3. :tada:

Also, see #4 for an area for future improvement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,141 @@
...
+**Every time a new message is input** through the message input box, we want to **save it to an array** (`messageStore`). Changes made to this array will be available to `<message>`
++ Create empty array to hold new messages (`messageStore`)
+
+renderMessage() is called once for every message (both from Redis and from Socket.io), so this is the ideal place to update our `messageStore` array.
+
++ Add messages to our array using push in `renderMessage()`
++ Pass the `messageStore` array to our riot `tag` as an option in riot.mount: `riot.mount('message', {messageStore: messageStore});`
++ `messageStore` is now available to be used in `<message>` and you can now replace the dummy data with the live data from this array using `this.messageStore = opts.messageStore`
++ Switch your data source from `dummyData` to `messageStore` in the riot `each` loop
+_Note: Any messages input into the message input box at this stage and going into the database and into our `messageStore` array, but are not yet being rendered on the page by riot - if you refresh your page however, messages are loaded in from Redis and will appear in the browser_.
```
Comment (jackpandas) [inline README.md]: Line 114 doesn't line-break properly from previous line

File: README.md [hunk 2]
```diff
@@ -1,4 +1,141 @@
-# Hapi.js Socket.io Redis Chat Example {Work in Progress}
+# Hapi.js Socket.io Redis Riot.js Chat Example [Work in Progress]
+
+This example adds [Riot.js](http://muut.com/riotjs/) to our popular [Hapi.js Socket.io and Redis Publish/Subscribe chat application example](https://github.com/dwyl/hapi-socketio-redis-chat-example), as per issue [#20]( https://github.com/dwyl/hapi-socketio-redis-chat-example/issues/20).
+
+The commits have been added in step-by-step so it should be simple to follow along. Watch this space!
+
++ [ ] Add links to riot documentation throughout & encourage pre-reading
```
Comment (minaorangina) [inline README.md]: riot --> Riot (keeping consistent with other JS frameworks mentioned in the README)

File: README.md [hunk 3]
```diff
@@ -1,4 +1,141 @@
...
+**Why is this happening?** 
++ Try adding `console.log($('#messages').height())` to `scrollToBottom()` and refresh your page to see what happens
+
+We're calling `scrollToBottom()` straight after calling `riot.mount` which means **Riot has not yet had the time to update the page**. What we want to do is call it right after `riot.mount` **_finishes_**. To fix this:
++ Pass the `scrollToBottom()` function as part of the options object so it's available to `<message>`
++ `this.on('mount')` tells Riot to perform an action once `mount` is complete, so add `this.on('mount', opts.scrollToBottom)` to `message.tag`.
+
+_Note: In step 4, we moved functions into our `message.tag` file to give Riot access to them. In this case, `scrollToBottom()` is being used in various places in `client.js` so we can't remove it completely otherwise `client.js` would lose access to it.    
+We could move `window.onresize` to `message.tag` as `window` is a global variable, but this is a bad [separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns).
+_
```
Comment (jackpandas) [inline README.md]: Underscore on line 137 breaks the emphasis?

File: README.md [hunk 4]
```diff
@@ -1,4 +1,141 @@
...
+  $('#messages').append(html);  // append to list
+  return;
+}
+```
+<br/>
+
+##Add Riot to the app & get it to display simple dummy data
+
+### 1. [Adds Riot.js 'Hello World'](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/13742162d8894e49684a6d27dcf1ea65f122180c) 
+Our first task is to add our [custom riot tag](https://muut.com/riotjs/guide/) into `index.html` and have it display some content (in this case, "Hello World"). This ensures we have riot set up properly.
```
Comment (minaorangina) [inline README.md]: riot --> Riot (two instances)

File: README.md [hunk 5]
```diff
@@ -1,4 +1,141 @@
...
++ Add the [Riot CDN](https://muut.com/riotjs/download.html) in a script tag at the bottom of your `index.html` (use the version which includes the _compiler_) 
++ Add your custom element (`<message></message>`) to `index.html`
++ Create your `.tag` file (where your _riot_ will live, in this case `message.tag`) and link it to `index.html` using a script tag (don't forget to add `type='riot/tag'`)
```
Comment (tsop14) [inline README.md]: Worth saying that .tag is a riot file extension with special syntax?
(Users shouldn't worry that text editors might not have syntax highlighting.)

File: README.md [hunk 6]
```diff
@@ -1,4 +1,141 @@
...
+
+```html
+<message>
+  <p>Hello World!</p>
+</message>
+```
+
+
+### 2. [Adds dummy array of data & iterates through it with Riot](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/9d8d80ab628bd71282a0eaf66ebb343d0358ca0c)
+We're using riot to replace dynamically generated HTML so we need to make sure our Riot can do the same job. We will **create some dummy data and render it with Riot**.
```
Comment (minaorangina) [inline README.md]: riot --> Riot

File: README.md [hunk 7]
```diff
@@ -1,4 +1,141 @@
...
++ Create some dummy data in the form of an array
+  + This will be in your `<message>` element within `message.tag` and should be contained within `<script>` tags for readability (though this isn't strictly necessary)
+  + Remember to attach your array to `this` so that riot recognises it
```
Comment (minaorangina) [inline README.md]: riot --> Riot (same for line 49)

File: README.md [hunk 8]
```diff
@@ -1,4 +1,141 @@
...
++ Use [riot's `each` functionality](https://muut.com/riotjs/guide/#loops) to iterate through that array and display the data in each `<li>`
+  + Pass your array of dummy data to the `each` loop
+  + The tag that you add the `each` attribute to (in this case, `<li>`) is the tag that will be repeated for every element of the array
+  + Note that properties of the array's elements (such as `{msg}` in this commit) can be referenced directly by name 
+  
+As always, you can [look at our commit](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/blob/9d8d80ab628bd71282a0eaf66ebb343d0358ca0c/public/message.tag) to see what we have added in this step.
+
+ 
+### 3. [Changes dummy data format & HTML structure in message.tag](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/88d53cc1601b564cfb6df695b1ccff0527558f1e)
+We want to ensure that our riot code will work with the structure of the data coming from our Redis database.
```
Comment (minaorangina) [inline README.md]: riot --> Riot

File: README.md [hunk 9]
```diff
@@ -1,4 +1,141 @@
...
+  s = this.leadZero(t.getSeconds());
+  return '' + h  + ':' + m + ':' + s;
+};
+``` 
++ **Call the `getTime()` function on our time (`t`) data**
+  + We need to use `parent.getTime()`. Riot would read `getTime()` as `this.getTime()` and `this` refers to the current element of the array as opposed to `<message>` which is where the function is accessible    
+
++ Delete the section of `renderMessage()` that appends the HTML to the `#messages` ul, so we understand what is being done by Riot versus our original code  
+  
+_ Note: At this point, if you try to type something into the message input box, it won't show up on the page as we've now removed the functionality of `renderMessage()` but have yet to replace it with riot functionality_
```
Comment (minaorangina) [inline README.md]: riot --> Riot

File: README.md [hunk 10]
```diff
@@ -1,4 +1,141 @@
...
+}
+```
+<br/>
+
+##Add Riot to the app & get it to display simple dummy data
+
+### 1. [Adds Riot.js 'Hello World'](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/13742162d8894e49684a6d27dcf1ea65f122180c) 
+Our first task is to add our [custom riot tag](https://muut.com/riotjs/guide/) into `index.html` and have it display some content (in this case, "Hello World"). This ensures we have riot set up properly.
++ Add the [Riot CDN](https://muut.com/riotjs/download.html) in a script tag at the bottom of your `index.html` (use the version which includes the _compiler_) 
++ Add your custom element (`<message></message>`) to `index.html`
```
Comment (jackpandas) [inline README.md]: Add your custom element to body of index.html

File: README.md (hunk 6 above)
Comment (tsop14) [inline README.md]: Is there a difference between 'riot' and 'Riot'? The above sentence could be read as such... If not let's pick one to use throughout?

File: README.md [hunk 11]
```diff
@@ -1,4 +1,141 @@
...
+  
+_ Note: At this point, if you try to type something into the message input box, it won't show up on the page as we've now removed the functionality of `renderMessage()` but have yet to replace it with riot functionality_
+
+<br/>
+
+##Hook up Riot to the live data from Redis
+### 5. [Change where we mount Riot](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/6cd72e0e4e18beffe8574847d91a7da1677e2988)
+We know that we only want to [mount](https://muut.com/riotjs/guide/#mounting) the `<message>` tag when there is data to display.
+
++ Move `riot.mount` to the `$.get()` callback within `loadMessages()` (in `client.js`) to ensure the riot element is mounted when there is data
```
Comment (minaorangina) [inline README.md]: riot --> Riot (but only where it says "...to ensure the riot element is...")

File: README.md [hunk 12]
```diff
@@ -1,4 +1,141 @@
...
+
+
+### 6. [Get data from Redis to be rendered using Riot](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/b755c52189805c0a402753cefecc2ae069a21383)
+**Every time a new message is input** through the message input box, we want to **save it to an array** (`messageStore`). Changes made to this array will be available to `<message>`
++ Create empty array to hold new messages (`messageStore`)
+
+renderMessage() is called once for every message (both from Redis and from Socket.io), so this is the ideal place to update our `messageStore` array.
+
++ Add messages to our array using push in `renderMessage()`
++ Pass the `messageStore` array to our riot `tag` as an option in riot.mount: `riot.mount('message', {messageStore: messageStore});`
```
Comment (minaorangina) [inline README.md]: array to our riot `tag` --> array to our Riot `tag`

File: README.md [hunk 13]
```diff
@@ -1,4 +1,141 @@
...
++ `messageStore` is now available to be used in `<message>` and you can now replace the dummy data with the live data from this array using `this.messageStore = opts.messageStore`
++ Switch your data source from `dummyData` to `messageStore` in the riot `each` loop
```
Comment (minaorangina) [inline README.md]: "...in the riot `each` loop" --> "...in the Riot `each` loop"

File: README.md [hunk 14]
```diff
@@ -1,4 +1,141 @@
...
+_Note: Any messages input into the message input box at this stage and going into the database and into our `messageStore` array, but are not yet being rendered on the page by riot - if you refresh your page however, messages are loaded in from Redis and will appear in the browser_.
+
+
+### 7. [Get riot to render new messages to the page](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/a71edd3da901ef6b01f5206578ffbb26aa810bdd)
```
Comment (minaorangina) [inline README.md]: "Get riot" --> "Get Riot"

File: README.md (hunk 6 above)
Comment (iteles) [inline README.md]: Agree!

File: README.md [hunk 15]
```diff
@@ -1,4 +1,141 @@
...
+:tada:  :tada:  :tada:  :tada: 
+##Bug fixes
+### 8. [scrollToBottom bug](https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/tree/b85121e0aa14c803d4041dd730b8189c86feab5f)
+It looks like someone has raised an issue on our repo!
+https://github.com/dwyl/hapi-socketio-redis-riot-chat-example/issues/3
+
+**Why is this happening?** 
++ Try adding `console.log($('#messages').height())` to `scrollToBottom()` and refresh your page to see what happens
+
+We're calling `scrollToBottom()` straight after calling `riot.mount` which means **Riot has not yet had the time to update the page**. What we want to do is call it right after `riot.mount` **_finishes_**. To fix this:
```
Comment (tsop14) [inline README.md]: Worth making this explanation more explicit?

...to update the page and there is nothing to scroll to the botom of!

File: README.md (hunk 5 above)
Comment (iteles) [inline README.md]: Definitely, adding it in :+1:

File: README.md (hunk 3 above)
Comment (iteles) [inline README.md]: Thanks! :+1:

File: README.md (hunk 15 above)
Comment (iteles) [inline README.md]: Done! :+1:

File: README.md (hunk 2 above)
Comment (iteles) [inline README.md]: Thanks for catching these @minaorangina :blush:

**Comment by iteles** [comment]
Thanks for reviewing @nelsonic! :+1:

**Comment by nelsonic** [comment]
you're awesome
Thanks @minaorangina @tsop14 @beechware @jackpandas @rjmk & @iteles

**Pull request merged**